addon/components/base/bs-dropdown.js

```diff
@@ -272,61 +283,88 @@ export default class Dropdown extends Component {
   @action
   toggleDropdown() {
     if (this.get('isOpen')) {
-      this.send('closeDropdown');
+      this.closeDropdown();
     } else {
-      this.send('openDropdown');
+      this.openDropdown();
     }
   }
 
   @action
   openDropdown() {
     this.set('isOpen', true);
-    this.addClickListener();
     this.get('onShow')();
   }
 
   @action
   closeDropdown() {
     this.set('isOpen', false);
-    this.removeClickListener();
     this.get('onHide')();
   }
 
-  addClickListener() {
-    if (!this.clickListener) {
-      this.clickListener = bind(this, this.closeOnClickHandler);
-      document.addEventListener('click', this.clickListener, true);
-    }
-  }
-
-  removeClickListener() {
-    if (this.clickListener)  {
-      document.removeEventListener('click', this.clickListener, true);
-      this.clickListener = null;
-    }
-  }
-
-  willDestroyElement() {
-    super.willDestroyElement(...arguments);
-    this.removeClickListener();
-  }
-
   /**
    * Handler for click events to close the dropdown
    *
    * @method closeOnClickHandler
    * @param e
    * @protected
    */
-  closeOnClickHandler(e) {
+  @action
+  closeHandler(e) {
     let { target } = e;
     let { toggleElement, menuElement } = this.getProperties('toggleElement', 'menuElement');
 
     if (!this.get('isDestroyed')
-      && (toggleElement && !toggleElement.contains(target))
-      && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick'))) {
-      this.send('closeDropdown');
+      && (
+        (e.type === 'keyup' && e.which === TAB_KEYCODE && menuElement && !menuElement.contains(target))
+        || (e.type === 'click' && toggleElement && !toggleElement.contains(target) && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick')))
+      )) {
+      this.closeDropdown();
+    }
+  }
+
+  @action
+  handleKeyEvent(event) {
+    if (/input|textarea/i.test(event.target.tagName)
+      ? (
+        event.which === SPACE_KEYCODE
+        || event.which !== ESCAPE_KEYCODE && (event.which !== ARROW_DOWN_KEYCODE && event.which !== ARROW_UP_KEYCODE || this.get('menuElement').contains(event.target)))
+      : !SUPPORTED_KEYCODES.includes(event.which)) {
+      return;
     }
+
+    event.preventDefault();
+    event.stopPropagation();
```

Comment:
Is this introducing a breaking change?

Reply:
Yeah, in theory *could* be. But tbh I don't see a way to introduce this *feature* without excluding the possibility to break things for some rare edge cases. Don't know how we should treat this. But honestly if we are very strict about it, anything we introduce could cause things to break in some more or less odd cases. Think of introducing a new private property in a component, but the user has extended from it already and added something with the same name. 🤔

In this case for example opening the dropdown by pressing the space key when the toggle is focused requires you to intercept the event, otherwise the default behaviour of the space key to scroll down the page would happen, which is not what you want here... (some for down key etc.)

Original bootstrap.js does the same (followed their implementation closely), so it is the right thing to do. But yes, could break things. But again I would consider this not a breaking change of our API, but a feature that introduces (potentially breaking) new behaviour. Idk if that's a valid perspective from the point of SemVer, but I think it is a valid one from a practical standpoint!? 😆

Reply:
Agree. But I think we should mention in release notes to give a hint for consumers that are affected by it.